volatility3/framework/symbols/linux/extensions/__init__.py

```diff
@@ -1897,3 +1960,250 @@ def get_file_mode(self) -> str:
+
+        Yields:
+            The inode's cached pages
+        """
+        if not self.i_size:
+            return
+        elif not (self.i_mapping and self.i_mapping.nrpages > 0):
+            return
+
+        page_cache = linux.PageCache(
+            context=self._context,
+            kernel_module_name="kernel",
+            page_cache=self.i_mapping.dereference(),
+        )
+        yield from page_cache.get_cached_pages()
+
+    def get_contents(self):
+        """Get the inode cached pages from the page cache
+
+        Yields:
+            page_index (int): The page index in the Tree. File offset is page_index * PAGE_SIZE.
+            page_content (str): The page content
+        """
+        for page_obj in self.get_pages():
+            page_index = int(page_obj.index)
+            page_content = page_obj.get_content()
+            yield page_index, page_content
+
+
+class address_space(objects.StructType):
+    @property
+    def i_pages(self):
+        """Returns the appropriate member containing the page cache tree"""
+        if self.has_member("i_pages"):
+            # Kernel >= 4.17
+            return self.member("i_pages")
+        elif self.has_member("page_tree"):
+            # Kernel < 4.17
+            return self.member("page_tree")
+
+        raise exceptions.VolatilityException("Unsupported page cache tree")
+
+
+class page(objects.StructType):
+    @property
+    @functools.lru_cache()
+    def pageflags_enum(self) -> Dict:
+        """Returns 'pageflags' enumeration key/values
+
+        Returns:
+            A dictionary with the pageflags enumeration key/values
+        """
+        # FIXME: It would be even better to use @functools.cached_property instead,
+        # however, this requires Python +3.8
+        try:
+            pageflags_enum = self._context.symbol_space.get_enumeration(
+                self.get_symbol_table_name() + constants.BANG + "pageflags"
+            ).choices
+        except exceptions.SymbolError:
+            vollog.debug(
+                "Unable to find pageflags enum. This can happen in kernels < 2.6.26 or wrong ISF"
+            )
+            # set to empty dict to show that the enum was not found, and so shouldn't be searched for again
+            pageflags_enum = {}
+
+        return pageflags_enum
+
+    def get_flags_list(self) -> List[str]:
+        """Returns a list of page flags
+
+        Returns:
+            List of page flags
+        """
+        flags = []
+        for name, value in self.pageflags_enum.items():
+            if self.flags & (1 << value) != 0:
+                flags.append(name)
+
+        return flags
+
+    def to_paddr(self) -> int:
+        """Converts a page's virtual address to its physical address using the current physical memory model.
+
+        Returns:
+            int: page physical address
+        """
+        vmlinux = linux.LinuxUtilities.get_module_from_volobj_type(self._context, self)
```

Comment:
This doesn't check the version of `LinuxUtilities`, so we need to be a little careful using it.  Since these are both core components it should be ok, but it might be better to just check the version and throw an exception is it's not right (you can check the `VersionRequirement` code to see how to do that).

Reply:
hm done, but I'm not sure we want this, have a look at the last commit

Reply:
Yeah, I see what you mean, but... I would like the check in there?  It looks like it could be reduced to a static method somewhere else that then appears as a oneliner at call sites, if the length/duplicative nature is what feels ugly?  I imagine a check like that will be useful for other code (or maybe even some that's already snuck in).

Reply:
Yeah, but it's becoming a checking versions infinite loop hehe. The get_module_from_volobj_type() was in LinuxUtilities to allow being reused, which need a version. Now, if we make another method to allow reusing the code to check the version of LinuxUtilities when calling get_module_from_volobj_type(), it won't also need a version :) ?

Reply:
@ikelos If I didn't miss anything, this should be the only open discussion remaining. Let me know if you're happy with how I fixed it

Reply:
I might've missed where this code was defined.  If it's defined *in* LinuxUtilities then it can be certain of its own version number, but otherwise it should be checking to see that it's available.  I didn't mean add another method to LinuxUtilities, I meant add it to a shareable base class, but it's fine if not.  I can sometimes try to reduce co-reuse too much...  5;P

Reply:
See my comment below, since these are changes to the linux extensions, they're rolled into the main framework version number, whose MINOR number should be bumped whenever an addition is made...

Reply:
(That also means we can rip out those LinuxUtilities checks I asked for and just rely on the framework and LinuxUtilities versions agreeing (sorry for the faffing, but getting these right so that people's plugins don't mysteriously not work on a unknown configurations is getting to be complex).